dace/transformation/interstate/state_fusion_with_happens_before.py

```diff
@@ -0,0 +1,589 @@
+            if freesyms and any(n.data in freesyms for n in first_state.nodes()
+                                if isinstance(n, nodes.AccessNode) and first_state.in_degree(n) > 0):
+                return False
+            # Fail if symbols assigned on the first edge are free symbols on the
+            # second edge
+            symbols_used = set(out_edges[0].data.free_symbols)
+            for e in in_edges:
+                if e.data.assignments.keys() & symbols_used:
+                    return False
+                # Also fail in the inverse; symbols assigned on the second edge are free symbols on the first edge
+                if new_assignments & set(e.data.free_symbols):
+                    return False
+
+        # There can be no state that have output edges pointing to both the
+        # first and the second state. Such a case will produce a multi-graph.
+        for src, _, _ in in_edges:
+            for _, dst, _ in graph.out_edges(src):
+                if dst == second_state:
+                    return False
+
+        if not permissive:
+            # Strict mode that inhibits state fusion if Python callbacks are involved
+            if Config.get_bool('frontend', 'dont_fuse_callbacks'):
+                for node in (first_state.data_nodes() + second_state.data_nodes()):
+                    if node.data == '__pystate':
+                        return False
+
+            # NOTE: This is quick fix for MPI Waitall (probably also needed for
+            # Wait), until we have a better SDFG representation of the buffer
+            # dependencies.
+            try:
+                next(node for node in first_state.nodes()
+                     if (isinstance(node, nodes.LibraryNode) and type(node).__name__ == 'Waitall')
+                     or node.label == '_Waitall_')
+                return False
+            except StopIteration:
+                pass
+            try:
+                next(node for node in second_state.nodes()
+                     if (isinstance(node, nodes.LibraryNode) and type(node).__name__ == 'Waitall')
+                     or node.label == '_Waitall_')
+                return False
+            except StopIteration:
+                pass
+
+            # If second state has other input edges, there might be issues
+            # Exceptions are when none of the states contain dataflow, unless
+            # the first state is an initial state (in which case the new initial
+            # state would be ambiguous).
+            first_in_edges = graph.in_edges(first_state)
+            second_in_edges = graph.in_edges(second_state)
+            if ((not second_state.is_empty() or not first_state.is_empty() or len(first_in_edges) == 0)
+                    and len(second_in_edges) != 1):
+                return False
+
+            # Get connected components.
+            first_cc = [cc_nodes for cc_nodes in nx.weakly_connected_components(first_state._nx)]
```

Comment:
The can_be_applied is quite computationally expensive for large graphs, how does this hold up in our pipeline for a decently sized application? I see it having a noticeable impact for graphs with larger states. Do you have any measurements / data?

Reply:
It's just as expensive as the baseline state fusion (and it's currently not part of the pipeline) - so unless state fusion itself is too slow, this should be ok.